lib/src/streams/stream_handler.dart

```diff
@@ -330,6 +330,17 @@ class StreamHandler extends Object with TerminatableMixin, ClosableMixin {
 
     _setupOutgoingMessageHandling(stream);
 
+    // handle incoming stream cancellation. RST is only sent when streamQueueOut
+    // has been closed because RST make the stream 'closed'.
+    streamQueueIn.onCancel.then((_) {
+      if (streamQueueIn.wasCancelled &&
+          !streamQueueIn.wasClosed &&
+          !streamQueueIn.wasTerminated &&
+          (streamQueueOut.wasClosed || streamQueueOut.wasTerminated)) {
+        _frameWriter.writeRstStreamFrame(streamId, ErrorCode.CANCEL);
```

Comment:
Unfortunately checking `streamQueueOut.wasClosed` is not sufficient to know that all data has been written out yet. There is a buffer in the `StreamQueueOut` as well as in the more low-level `ConnectionMessageQueueOut`: Both of them buffer data until the flow control window allows us to send the data to the underlying socket.

There is one way we can ensure the ordering, which is by adding a new `Message` type. We send it down to `StreamQueueOut` and then to `ConnectionMessageQueueOut`, which will then write the RST frame.

```dart
streamQueueIn.onCancel.then((_) {
  // If our side is done sending data, i.e. we have enqueued the end-of-steam in the
  // outgoing message queue, but the remote end is still sending us data, despite us not
  // being interested in it, we will reset the stream.
  if (stream.state == StreamState.HalfClosedLocal) {
    stream.outgoingQueue.enqueueMessage(new ResetStreamMessage(stream.id, ErrorCode.CANCEL));`
  }
});
```

(similar to how we do it with `GoawayMessage`)

That should work. Unfortunately the server will also need some fixing in this scenario, because when it receives the RST it will immediately tell the application, not delivering the data that might be in the connection/stream level buffers.


To make life easier for you, I'm perfectly fine if you don't handle the RST case for now and just drop the messages.

Reply:
Making this change the _Bad state_ error is back because we try to enqueue after closing :-(

```
ERROR: Bad state: Was in the process of closing.
package:http2/src/error_handler.dart 91:7              _StreamMessageQueueOut&Object&TerminatableMixin&ClosableMixin.ensureNotClosingSync
package:http2/src/flowcontrol/stream_queues.dart 71:5  StreamMessageQueueOut.enqueueMessage
package:http2/src/streams/stream_handler.dart 341:30   StreamHandler._newStreamInternal.<fn>
```

Reply:
I see this error in the test I added `data-frame-received-after-stream-cancel`

Reply:
>I see this error in the test I added data-frame-received-after-stream-cancel

 and also in a simple manual test I wrote

Reply:
According to https://http2.github.io/http2-spec/#StreamStates it is allowed to send R after ES. I will update the code to allow that.

Reply:
Done